File: code-monkey/src/code_context.py
import os
import shutil
import logging
from typing import List
import pathspec
from constants import get_artifacts_dir, get_root_dir

def get_all_src_files() -> List[str]:
    src_files = []

    root_dir = get_root_dir()

    # Read .gitignore patterns
    gitignore_patterns = [".git"]
    # Check current directory and up to max 3 parent directories until we hit
    # our root dir.
    for i in range(4):
        dir_path = os.path.abspath(os.path.join(os.path.dirname(__file__), *[".." for _ in range(i)]))

        gitignore_path = os.path.join(dir_path, ".gitignore")
        logging.debug(f"checking gitignore path {gitignore_path}")
        if os.path.exists(gitignore_path):
            logging.debug("   found it")
            with open(gitignore_path, 'r') as gitignore_file:
                gitignore_patterns.extend(gitignore_file.read().splitlines())

        if (dir_path == root_dir):
            break

    # Create PathSpec object
    ignore_spec = pathspec.GitIgnoreSpec.from_lines(gitignore_patterns)

    for root, dirs, files in os.walk(get_root_dir()):
        for file in files: 
            src_path = os.path.join(root, file)
            rel_path = os.path.relpath(src_path, get_root_dir())
            if not ignore_spec.match_file(rel_path):
                src_files.append(rel_path)
    return src_files



class CodeContext:
    def __init__(self):
        self.known_files = []

    def copy_src(self) -> List[str]:
        dest_dir = get_artifacts_dir()

        if not os.path.exists(dest_dir):
            os.makedirs(dest_dir)

        files_to_copy = get_all_src_files()

        for rel_path in files_to_copy:
            src_path = os.path.join(get_root_dir(), rel_path)
            dest_path = os.path.join(dest_dir, rel_path)
            os.makedirs(os.path.dirname(dest_path), exist_ok=True)
            shutil.copy2(src_path, dest_path)

        self.known_files = files_to_copy
        return self.known_files

if __name__ == "__main__":
    for file in get_all_src_files():
        print(file)

File: code-monkey-ts/src/code_context.ts
import * as fs from 'fs';
import * as path from 'path';
import * as util from 'util';
import { getArtifactsDir, getRootDir } from './constants';

const readFile = util.promisify(fs.readFile);
const writeFile = util.promisify(fs.writeFile);
const mkdir = util.promisify(fs.mkdir);
const copyFile = util.promisify(fs.copyFile);

// Simple implementation of gitignore-style pattern matching
function matchesGitIgnorePattern(filePath: string, pattern: string): boolean {
    const regexPattern = pattern
        .replace(/\./g, '\\.')
        .replace(/\*/g, '.*')
        .replace(/\?/g, '.');
    const regex = new RegExp(`^${regexPattern}$`);
    return regex.test(filePath);
}

async function getAllSrcFiles(): Promise<string[]> {
    const srcFiles: string[] = [];
    const rootDir = getRootDir();

    // Read .gitignore patterns
    let gitignorePatterns: string[] = ['.git'];

    // Check current directory and up to max 3 parent directories until we hit our root dir
    for (let i = 0; i < 4; i++) {
        const dirPath = path.resolve(__dirname, ...Array(i).fill('..'));
        const gitignorePath = path.join(dirPath, '.gitignore');

        console.debug(`checking gitignore path ${gitignorePath}`);
        if (fs.existsSync(gitignorePath)) {
            console.debug('   found it');
            const gitignoreContent = await readFile(gitignorePath, 'utf-8');
            gitignorePatterns = gitignorePatterns.concat(gitignoreContent.split('\n'));
        }

        if (dirPath === rootDir) {
            break;
        }
    }

    // Walk through the directory
    const walk = (dir: string): void => {
        const files = fs.readdirSync(dir);
        for (const file of files) {
            const filePath = path.join(dir, file);
            const stat = fs.statSync(filePath);
            if (stat.isDirectory()) {
                walk(filePath);
            } else {
                const relPath = path.relative(rootDir, filePath);
                if (!gitignorePatterns.some(pattern => matchesGitIgnorePattern(relPath, pattern))) {
                    srcFiles.push(relPath);
                }
            }
        }
    };

    walk(rootDir);
    return srcFiles;
}

class CodeContext {
    private knownFiles: string[] = [];

    async copySrc(): Promise<string[]> {
        const destDir = getArtifactsDir();

        if (!fs.existsSync(destDir)) {
            await mkdir(destDir, { recursive: true });
        }

        const filesToCopy = await getAllSrcFiles();

        for (const relPath of filesToCopy) {
            const srcPath = path.join(getRootDir(), relPath);
            const destPath = path.join(destDir, relPath);
            await mkdir(path.dirname(destPath), { recursive: true });
            await copyFile(srcPath, destPath);
        }

        this.knownFiles = filesToCopy;
        return this.knownFiles;
    }
}

// Main execution
if (require.main === module) {
    (async () => {
        const files = await getAllSrcFiles();
        for (const file of files) {
            console.log(file);
        }
    })();
}

export { CodeContext, getAllSrcFiles };
